lib/galaxy/webapps/galaxy/api/file_sources.py

```diff
@@ -0,0 +1,107 @@
+)
+
+
+@router.cbv
+class FastAPIFileSources:
+    file_source_instances_manager: FileSourceInstancesManager = depends(FileSourceInstancesManager)
+
+    @router.get(
+        "/api/file_source_templates",
+        summary="Get a list of file source templates available to build user defined file sources from",
+        response_description="A list of the configured file source templates.",
+        operation_id="file_sources__templates_index",
+    )
+    def index_templates(
+        self,
+        trans: ProvidesUserContext = DependsOnTrans,
+    ) -> FileSourceTemplateSummaries:
+        return self.file_source_instances_manager.summaries
+
+    @router.post(
+        "/api/file_source_instances",
+        summary="Create a user-bound object store.",
+        operation_id="file_sources__create_instance",
+    )
+    def create(
+        self,
+        trans: ProvidesUserContext = DependsOnTrans,
+        payload: CreateInstancePayload = Body(...),
+    ) -> UserFileSourceModel:
+        return self.file_source_instances_manager.create_instance(trans, payload)
+
+    @router.get(
+        "/api/file_source_instances",
+        summary="Get a list of persisted file source instances defined by the requesting user.",
+        operation_id="file_sources__instances_index",
+    )
+    def instance_index(
+        self,
+        trans: ProvidesUserContext = DependsOnTrans,
+    ) -> List[UserFileSourceModel]:
+        return self.file_source_instances_manager.index(trans)
+
+    @router.get(
+        "/api/file_source_instances/{user_file_source_id}",
+        summary="Get a list of persisted file source instances defined by the requesting user.",
+        operation_id="file_sources__instances_get",
+    )
+    def instances_show(
+        self,
+        trans: ProvidesUserContext = DependsOnTrans,
+        user_file_source_id: str = UserFileSourceIdPathParam,
+    ) -> UserFileSourceModel:
+        return self.file_source_instances_manager.show(trans, user_file_source_id)
+
+    @router.put(
+        "/api/file_source_instances/{user_file_source_id}",
+        summary="Update or upgrade user file source instance.",
+        operation_id="file_sources__instances_update",
+    )
+    def update_instance(
+        self,
+        trans: ProvidesUserContext = DependsOnTrans,
+        user_file_source_id: str = UserFileSourceIdPathParam,
+        payload: ModifyInstancePayload = Body(...),
+    ) -> UserFileSourceModel:
+        return self.file_source_instances_manager.modify_instance(trans, user_file_source_id, payload)
+
+    @router.delete(
+        "/api/file_source_instances/{user_file_source_id}",
+        summary="Purge user file source instance.",
+        operation_id="file_sources__instances_purge",
+    )
+    def purge_instance(
+        self,
+        trans: ProvidesUserContext = DependsOnTrans,
+        user_file_source_id: str = UserFileSourceIdPathParam,
+    ) -> None:
```

Comment:
Consider returning 204 instead of just None as it will be the same as returning `null`. It is not a big deal but it feels a bit more standard.
The same for the equivalent for object stores :)

```diff
--- a/lib/galaxy/webapps/galaxy/api/file_sources.py
+++ b/lib/galaxy/webapps/galaxy/api/file_sources.py
@@ -4,6 +4,8 @@ from typing import List
 from fastapi import (
     Body,
     Path,
+    Response,
+    status,
 )
 
 from galaxy.files.templates import FileSourceTemplateSummaries
@@ -98,10 +100,12 @@ class FastAPIFileSources:
         "/api/file_source_instances/{user_file_source_id}",
         summary="Purge user file source instance.",
         operation_id="file_sources__instances_purge",
+        status_code=status.HTTP_204_NO_CONTENT,
     )
     def purge_instance(
         self,
         trans: ProvidesUserContext = DependsOnTrans,
         user_file_source_id: str = UserFileSourceIdPathParam,
-    ) -> None:
+    ):
         self.file_source_instances_manager.purge_instance(trans, user_file_source_id)
+        return Response(status_code=status.HTTP_204_NO_CONTENT)
```

Reply:
Done - thanks so much!